modules/ipp/scanner.go

```diff
@@ -126,86 +172,394 @@ func (scanner *Scanner) GetPort() uint {
-func getIPPURL(https bool, host string, port uint16, endpoint string) string {
-	var proto string
-	if https {
-		proto = "https"
-	} else {
-		proto = "http"
+func ippInContentType(resp http.Response) (bool, error) {
+	// TODO: ?See if capturing parameters gets anything interesting in 1% scan??
+	// Parameters can be ignored, since there are no required or optional parameters
+	// IPP parameters specified at https://www.iana.org/assignments/media-types/application/ipp
+	mediatype, _, err := mime.ParseMediaType(resp.Header.Get("Content-Type"))
+	// FIXME: See if empty media type is sufficient as failure indicator,
+	// there could be other states where reading mediatype screwed up, but isn't empty (ie: corrupted/malformed)
+	if mediatype == "" && err != nil {
+		//TODO: Handle errors in a weird way, since media type is still returned
+		//      if there's an error when parsing optional parameters
+		return false, err
 	}
-	return proto + "://" + host + ":" + strconv.FormatUint(uint64(port), 10) + endpoint
+	// FIXME: Maybe pass the error along, maybe not. We got what we wanted.
+	return mediatype == ContentType, nil
 }
 
-// FIXME: Ensure that an actual content type is being used with some library
-func ippInContentType(resp http.Response) bool {
-	return strings.Contains(resp.Header.Get("Content-Type"), ContentType)
+// FIXME: Cleaner to write this code, possibly slower than copy-pasted version
+// FIXME: Quite possibly not easier to read ("What does storeBody do? Where does it store it?")
+// FIXME: Add some error handling somewhere in here, unless errors should just be ignored and we get what we get
+func storeBody(res *http.Response, scanner *Scanner) {
+	b := bufferFromBody(res, scanner)
+	res.BodyText = b.String()
+	if len(res.BodyText) > 0 {
+		m := sha256.New()
+		m.Write(b.Bytes())
+		res.BodySHA256 = m.Sum(nil)
+	}
 }
 
-// FIXME: Change to instead return (*ScanResults, *zgrab2.ScanError)
-// TODO: Doesn't support TLS at all right now
-func (scanner *Scanner) grab(target zgrab2.ScanTarget) (int8, int8, *zgrab2.ScanError) {
-	//FIXME: This is not where this hostname assignment logic should live
-	host := target.Domain
-	if host == "" {
-		host = target.IP.String()
+// TODO: Make sure this it isn't too slow to use this instead of copy-pasting everywhere necessary
+func bufferFromBody(res *http.Response, scanner *Scanner) *bytes.Buffer {
+	b := new(bytes.Buffer)
+	maxReadLen := int64(scanner.config.MaxSize) * 1024
+	readLen := maxReadLen
+	if res.ContentLength >= 0 && res.ContentLength < maxReadLen {
+		readLen = res.ContentLength
+	}
+	io.CopyN(b, res.Body, readLen)
+	res.Body.Close()
+	res.Body = ioutil.NopCloser(b)
+	return b
+}
+
+// FIXME: add proper error handling to this
+// TODO: Support reading from multiple instances of the same attribute in a response
+func readAttributeFromBody(attrString string, body *[]byte) ([][]byte, error) {
+	attr := []byte(attrString)
+	interims := bytes.Split(*body, attr)
```

Comment:
Is this valid? Like, what if there was an attribute `max-supported-cups-version` -- that would get picked up by `readAttributeFromBody` wouldn't it?

Reply:
You're definitely right. My first thought here is to write a bare bones IPP response parser. Does that seem appropriate, or is there an alternative plan of action you'd suggest?